lgc/disassembler/Disassembler.cpp

```diff
@@ -0,0 +1,506 @@
+}
+
+// =====================================================================================================================
+// Disassemble one section.
+void ObjDisassembler::processSection(ELFSectionRef sectionRef) {
+  // Omit certain ELF sections.
+  unsigned sectType = sectionRef.getType();
+  if (sectType == ELF::SHT_NULL || sectType == ELF::SHT_STRTAB || sectType == ELF::SHT_SYMTAB ||
+      sectType == ELF::SHT_REL || sectType == ELF::SHT_RELA)
+    return;
+
+  m_streamer->AddBlankLine();
+  unsigned flags = sectionRef.getFlags();
+  bool isCode = flags & ELF::SHF_EXECINSTR;
+  MCSection *sect = m_context->getELFSection(cantFail(sectionRef.getName()), sectType, flags);
+  m_streamer->SwitchSection(sect);
+  unsigned instAlignment = 1;
+  if (isCode)
+    instAlignment = m_context->getAsmInfo()->getMinInstAlignment();
+
+  // Create an array of all symbols in this section. Also emit directives for symbol type and size.
+  // The size is an expression endSym-sym where endSym is a synthesized label at the end of the function.
+  std::vector<std::unique_ptr<std::string>> synthesizedLabels;
+  SectionSymbolsTy symbols;
+  for (ELFSymbolRef symbolRef : m_objFile->symbols()) {
+    if (cantFail(symbolRef.getSection()) != sectionRef)
+      continue;
+    symbols.push_back(
+        SymbolInfoTy(cantFail(symbolRef.getValue()), cantFail(symbolRef.getName()), symbolRef.getELFType()));
+
+    MCSymbol *sym = m_context->getOrCreateSymbol(symbols.back().Name);
+    switch (symbols.back().Type) {
+    case ELF::STT_FUNC:
+      m_streamer->emitSymbolAttribute(sym, MCSA_ELF_TypeFunction);
+      break;
+    case ELF::STT_OBJECT:
+      m_streamer->emitSymbolAttribute(sym, MCSA_ELF_TypeObject);
+      break;
+    }
+
+    if (uint64_t size = symbolRef.getSize()) {
+      if (symbols.back().Addr + size <= cantFail(sectionRef.getContents()).size()) {
+        synthesizedLabels.push_back(std::make_unique<std::string>((Twine(symbols.back().Name) + "_symend").str()));
+        StringRef endName = *synthesizedLabels.back();
+        MCSymbol *endSym = m_context->getOrCreateSymbol(endName);
+        symbols.push_back(SymbolInfoTy(symbols.back().Addr + size, endName, ELF::STT_NOTYPE));
+        const MCExpr *sizeExpr = MCBinaryExpr::createSub(MCSymbolRefExpr::create(endSym, *m_context),
+                                                         MCSymbolRefExpr::create(sym, *m_context), *m_context);
+        m_streamer->emitELFSize(sym, sizeExpr);
+      }
+    }
+  }
+
+  // Collect and sort the relocs for the section.
+  std::vector<object::RelocationRef> relocs;
+  for (auto &relSect : m_objFile->sections()) {
+    auto relocatedSectOr = relSect.getRelocatedSection();
+    if (relocatedSectOr && *relocatedSectOr == sectionRef) {
+      for (auto &reloc : relSect.relocations())
+        relocs.push_back(reloc);
+    }
+  }
+  std::sort(relocs.begin(), relocs.end(), [](const object::RelocationRef &a, const object::RelocationRef &b) {
```

Comment:
nit: use `llvm::sort`

Reply:
Until you pointed this out and looked it up, I didn't know you're supposed to do that to save code bloat from template expansion.